pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -1661,3 +1726,29 @@ func (ctrl *PersistentVolumeController) findDeletablePlugin(volume *v1.Persisten
 	}
 	return plugin, nil
 }
+
+// obtain provisioner/deleter name for a volume
+func (ctrl *PersistentVolumeController) getProvisionerNameFromVolume(volume *v1.PersistentVolume) string {
+	storageClass := v1helper.GetPersistentVolumeClass(volume)
+	class, err := ctrl.classLister.Get(storageClass)
+	if err != nil {
+		return "N/A"
+	}
+	return class.Provisioner
```

Comment:
this needs to have migration specific logic to return the correct provisioner name. Is there any reason why we're not using "pv.kubernetes.io/provisioned-by" annotation on the PV?

Reply:
thanks for clarifying. there was concern about "provisioned-by" could not be reliably used though my test seems showing that annotation is actually getting updated correctly by external provisioners. I am happy to switch to use "provisioned-by" annotation if its reliable. Does it also cover the in-tree case?

Reply:
I think the old code used to use `provisioned-by` implicitly through the return value of `findDeletablePlugin`. I can't comment on its reliability for general external provisioners. You're right that maybe it's safer to use the storage class provisioner name, lets just do a translation of that then

Reply:
@davidz627 yes indeed as you commented, the original logic to look for plugin/provisioner for deletion uses the annotation ([check here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1636)) ONLY if it exists. otherwise, it uses volume spec for finding. That was the major reason why the reliability concern was raised above. I will keep the implementation as its for now then. thnx

Reply:
I still think the storage class gotten here is going to be the in-tree one when migration is on. But I see we also dont have the plugin information to check whether this volume has been migrated. Could we use `provisioned-by` and fall back to the storage class provisioner if we can't find the `provisioned-by` annotation. (at least for migration cases this annotation should exist)

Reply:
I think this has to be fixed correctly. i.e - we should fetch the `DeletablePlugin` from Provisioner name retrieved from storageclass. Check if migration is enabled for the plugin and use CSI name otherwise use provisioner name present in SC.

Reply:
@msau42 also clarified that using `provisioned-by` could cause explosion of metric names for local-storage volumes because `provisioned-by` annotation points to node name in that case.

Reply:
@gnufied @davidz627 
instead of using storageclass defined provisioner name, to find the DeletablePlugin (if exists), I am proposing the following for getting the deleter name:
if has dynamic provisioned annotation, check the value of it, for external provisioner, use the value directly as the name, otherwise, use the plugin name to find DeletablePlugin and get the name.
If not dynamically provisioned, use spec to find plugin, falls back to "N/A" if none of the above.
WDYT?

Reply:
> @msau42 also clarified that using `provisioned-by` could cause explosion of metric names for local-storage volumes because `provisioned-by` annotation points to node name in that case.

"provision-by" is not used in this case for the concerns listed above